### DISCUSSION TITLE
Static classes with identical names and namespaces into different referenced projects

### DISCUSSION THREAD

**Discussion by CSharpLanguage** [dotnet/csharplang#1722] [Language Ideas]
### Scenario

Try to extend by new methods a existing static class from another referenced library (project) 

### Steps to reproduce

1. Create library Project A and add a static class with some methods

```c#
namespace AnyLibrary
{
    public static class AnyExtensions
    {
        public static T Method_A<T>(this T o) => o;

        public static T Method_B<T>(T o) => o;
    }
}
```

2. Create Project B and add into references Project A

3. Add into Project B a static class **with the same name and namespace but another methods**

```c#
namespace AnyLibrary
{
    public static class AnyExtensions
    {
        public static T Method_C<T>(this T o) => o;

        public static T Method_D<T>(T o) => o;
    }
}
```

4. Try to use all methods

```c#
using AnyLibrary;

var x = "abc";
Console.WriteLine(x.Method_A()); // Compiled !!! But show Error CS1061 if there are another fatal errors into project	 string' does not contain a definition for 'Method_A' and no extension method 'Method_A' accepting a first argument of type 'string' could be found (are you missing a using directive or an assembly reference?)
//Console.WriteLine(AnyExtensions.Method_A(x)); // Error	CS0117 'AnyExtensions' does not contain a definition for 'Method_A'
//Console.WriteLine(AnyExtensions.Method_B(x)); // Error	CS0117 'AnyExtensions' does not contain a definition for 'Method_B'

Console.WriteLine(x.Method_C()); // Compiled
Console.WriteLine(AnyExtensions.Method_C(x)); // Compiled
Console.WriteLine(AnyExtensions.Method_D(x)); // Compiled
```

5. Remarkable at a current situation that it is possible to use extension method like `x.Method_A()` but not possible like `AnyExtensions.Method_A(x)`. Seems that 
```c#
AnyExtensions.Method_A(x);
AnyExtensions.Method_B(x)
```
should be allowed too by compiler because it is possible to resolve a path conflict by method names.

**Comment by theunrepentantgeek** [comment]
[Using directives](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-directive) allow specifying an alias, explicitly to work around this issue.

I'd conjecture that the reason for method resolution *not* doing this disambiguation is because it would introduce inconsistency. If both methods declared `Method_Z()`, then only can be called - with your demo, I think it would be the second one but would have to test. Having both `Method_Z()` and `Method_A()` work at the same time, even though they're resident in different classes, would be very odd.

**Comment by jnm2** [comment]
Such a conflict is extremely undesirable. When it exists, switching back and forth in the same file should be minimized. A good design will not require [extern aliases](https://docs.microsoft.com/dotnet/csharp/language-reference/keywords/extern-alias).
(Aside: applying aliases to assemblies from NuGet package references requires painful hacks.)

**Comment by CSharpLanguage** [comment]
@tannergooding @jnm2 
Maybe described steps are not clear.

The real situation looks like:
1. There is a common library which had used into different solutions

2. The library contains a similar static class with some methods

```c#
namespace AnyLibrary
{
    public static class AnyExtensions
    {
        public static T Method_A<T>(this T o) => ... ;

        public static T Method_B<T>(T o) => ... ;

        public static T Method_C<T>(this T o) => ... ;

        public static T Method_D<T>(T o) => ... ;
    }
}
```

3. Implementation of methods C and D require external dependency. But this dependency is not important for some  solutions and may be avoided

4. By this reason I try to split AnyLibrary on two parts: the first contains

```c#
namespace AnyLibrary
{
    public static class AnyExtensions
    {
        public static T Method_A<T>(this T o) => ... ;

        public static T Method_B<T>(T o) => ... ;
    }
}
```

and not require the external dependency,
the second optional part
 
```c#
namespace AnyLibrary
{
    public static class AnyExtensions
    {
        public static T Method_C<T>(this T o) => ... ;

        public static T Method_D<T>(T o) => ... ;
    }
}
```

contains the external dependency

5. Now I expect that if both parts referenced to the any project I can use all following variations

```c#
x.Method_A()
AnyExtensions.Method_A(x)
AnyExtensions.Method_B(x)

x.Method_C()
AnyExtensions.Method_C(x)
AnyExtensions.Method_D(x)
```

But currently situation looks very strange

1. I can use extension methods like
```c#
x.Method_A()
x.Method_C()
```
but can't use **the same** methods via class name
```c#
AnyExtensions.Method_A(x)
AnyExtensions.Method_C(x)
```

2. If I try to define the second part inside the project and add the first part a like external dependency I can use 
```c#
AnyExtensions.Method_C(x)
AnyExtensions.Method_D(x)
```
but can't
```c#
AnyExtensions.Method_A(x)
AnyExtensions.Method_B(x)
```

If a compiler can resolve this path conflict for implicit extension methods call [`x.Method_A()`], why it can not resolve the same conflict for the explicit extension method call [`AnyExtensions.Method_A(x)`] and a simple static methods too [`AnyExtensions.Method_B(x)`] ?

**Comment by jnm2** [comment]
@CyrusNajmabadi Can we reopen when conversion to discussion fails so that it doesn't appear that the issue is being rejected without a response?

**Comment by CyrusNajmabadi** [comment]
![image](https://user-images.githubusercontent.com/4564579/96750259-9a46b980-1380-11eb-97f5-c26959a023ee.png)

:(
